### PR TITLE
[SPARK-21743][SQL] top-most limit should not cause memory leak

### PR DESCRIPTION
**Pull request by cloud-fan**
## What changes were proposed in this pull request?

For top-most limit, we will use a special operator to execute it: `CollectLimitExec`.

`CollectLimitExec` will retrieve `n`(which is the limit) rows from each partition of the child plan output, see https://github.com/apache/spark/blob/v2.2.0/sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala#L311. It's very likely that we don't exhaust the child plan output.

This is fine when whole-stage-codegen is off, as child plan will release the resource via task completion listener. However, when whole-stage codegen is on, the resource can only be released if all output is consumed.

To fix this memory leak, one simple approach is, when `CollectLimitExec` retrieve `n` rows from child plan output, child plan output should only have `n` rows, then the output is exhausted and resource is released. This can be done by wrapping child plan with `LocalLimit`

## How was this patch tested?

a regression test

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @gengliangwang @sameeragarwal @hvanhovell

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala [hunk 1]
```diff
@@ -2658,4 +2658,9 @@ class SQLQuerySuite extends QueryTest with SharedSQLContext {
       checkAnswer(sql("SELECT __auto_generated_subquery_name.i from (SELECT i FROM v)"), Row(1))
     }
   }
+
+  test("SPARK-21743: top-most limit should not cause memory leak") {
+    // In unit test, Spark will fail the query if memory leak detected.
```
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: The test did not fail, but I saw the warning message:
> 22:05:07.455 WARN org.apache.spark.executor.Executor: Managed memory leak detected; size = 33554432 bytes, TID = 2

Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: did you try this test in spark shell? We only throw exception for memory leak if `spark.unsafe.exceptionOnMemoryLeak` is true. But this config is false by default, and is true in unit test.

**Comment by cloud-fan** [comment]
also cc @viirya @kiszk

File: sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegenExec.scala [hunk 2]
```diff
@@ -474,6 +474,10 @@ case class CollapseCodegenStages(conf: SQLConf) extends Rule[SparkPlan] {
   }
 
   private def supportCodegen(plan: SparkPlan): Boolean = plan match {
+    // Do not enable whole stage codegen for a single limit.
+    case limit: BaseLimitExec if !limit.child.isInstanceOf[CodegenSupport] ||
+        !limit.child.asInstanceOf[CodegenSupport].supportCodegen =>
```
Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegenExec.scala]: We can also override `LocalLimitExec.supportCodegen` and make that depend on the `child.supportCodegen`. That seems more elegant that to special case it here.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala [hunk 3]
```diff
@@ -72,7 +72,12 @@ abstract class SparkStrategies extends QueryPlanner[SparkPlan] {
           execution.TakeOrderedAndProjectExec(
             limit, order, projectList, planLater(child)) :: Nil
         case logical.Limit(IntegerLiteral(limit), child) =>
-          execution.CollectLimitExec(limit, planLater(child)) :: Nil
+          // Normally wrapping child with `LocalLimitExec` here is a no-op, because
+          // `CollectLimitExec.executeCollect` will call `LogicalLimitExec.executeTake`, which
+          // calls `child.executeTake`. If child supports whole stage codegen, adding this
+          // `LocalLimitExec` can break the input consuming loop inside whole stage codegen and
```
Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala]: By `...break the input consuming loop...` you mean stop processing, and break whole stage code gen. We may need to word this slightly differently :)...

**Comment by hvanhovell** [comment]
LGTM - pending jenkins

File: sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala [hunk 4]
```diff
@@ -54,6 +54,16 @@ trait BaseLimitExec extends UnaryExecNode with CodegenSupport {
...
 
+  // Do not enable whole stage codegen for a single limit.
+  override def supportCodegen: Boolean = child match {
+    case plan: CodegenSupport => plan.supportCodegen
+    case _ => false
+  }
+
+  override def executeTake(n: Int): Array[InternalRow] = child.executeTake(math.min(n, limit))
+
+  override def executeCollect(): Array[InternalRow] = child.executeTake(limit)
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala]: `executeTake` looks good. But should `executeCollect` be the same for `LocalLimitExec` and `GlobalLimitExec`?

`doExecute` is an example. For `LocalLimitExec`, it takes `limit` rows in each partition. For `GlobalLimitExec`, it takes `limit` rows in single partition.

Previously `executeCollect` retrieves `limit` rows from each partition. After this change, `executeCollect` for `LocalLimitExec` retrieves only `limit` rows.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala [hunk 5]
```diff
@@ -72,7 +72,12 @@ abstract class SparkStrategies extends QueryPlanner[SparkPlan] {
           execution.TakeOrderedAndProjectExec(
             limit, order, projectList, planLater(child)) :: Nil
         case logical.Limit(IntegerLiteral(limit), child) =>
-          execution.CollectLimitExec(limit, planLater(child)) :: Nil
+          // Normally wrapping child with `LocalLimitExec` here is a no-op, because
+          // `CollectLimitExec.executeCollect` will call `LogicalLimitExec.executeTake`, which
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala]: typo? `LogicalLimitExec` -> `LocalLimitExec`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala (hunk 4 above)
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala]: Seems this fix relies `CollectLimitExec.executeCollect` to call `LocalLimitExec.executeTake`. Looks like we don't need to change `executeCollect`?

**Comment by viirya** [comment]
LGTM except for the question with `executeCollect`.

**Comment by dongjoon-hyun** [comment]
Retest this please.

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala (hunk 1 above)
Comment (kiszk) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: When this is also executed on Intellij, this test does not fail. How about this?
```
class SQLQuerySparkContextSuite extends QueryTest with LocalSparkContext {
  val spark = SparkSession
    .builder()
    .config("spark.unsafe.exceptionOnMemoryLeak", "true")
    .master("local[1]")
    .getOrCreate()
  test("SPARK-21743: top-most limit should not cause memory leak") {
    spark.range(100).groupBy("id").count().limit(1).collect()
  }
}
```

Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: That should be fine, as long as our test framework can capture it. : )

Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: This issue is fixed in https://github.com/apache/spark/pull/18967 

I think we need to move this test case to `DataFrameSuite`

Comment (sameeragarwal) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: How about just adding it in `TestSparkSession` instead?

**Comment by viirya** [comment]
LGTM

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM, too.

**Comment by gatorsmile** [comment]
Thanks! Merging to master.

**Pull request closed without merging**